partiql-eval/src/eval/evaluable.rs

```diff
@@ -282,6 +282,79 @@ impl Evaluable for EvalJoin {
     }
 }
 
+/// Represents an evaluation `GROUP BY` operator. For `GROUP BY` operational semantics, see section
+/// `11` of
+/// [PartiQL Specification — August 1, 2019](https://partiql.org/assets/PartiQL-Specification.pdf).
+#[derive(Debug)]
+pub struct EvalGroupBy {
+    pub strategy: EvalGroupingStrategy,
+    pub exprs: HashMap<String, Box<dyn EvalExpr>>,
+    pub group_as_alias: Option<String>,
+    pub input: Option<Value>,
+}
+
+/// Represents the grouping qualifier: ALL or PARTIAL.
+#[derive(Debug)]
+pub enum EvalGroupingStrategy {
+    GroupFull,
+    GroupPartial,
+}
+
+/// Represents the items in a `GROUP BY` list. e.g. `t.a AS a` in `... GROUP BY t.a AS a`.
+#[derive(Debug)]
+pub struct GroupKey {
+    pub expr: Box<dyn EvalExpr>,
+    pub as_alias: Option<Box<dyn EvalExpr>>,
+}
+
+impl Evaluable for EvalGroupBy {
+    fn evaluate(&mut self, ctx: &dyn EvalContext) -> Option<Value> {
+        let group_as_alias = &self.group_as_alias;
+        let input_value = self.input.take().expect("Error in retrieving input value");
+
+        match self.strategy {
+            EvalGroupingStrategy::GroupPartial => todo!(),
+            EvalGroupingStrategy::GroupFull => {
+                let mut groups: HashMap<Tuple, Vec<Value>> = HashMap::new();
+                for v in input_value.into_iter() {
+                    let v_as_tuple = v.coerce_to_tuple();
+                    let mut evaluated_values = Tuple::new();
+
+                    self.exprs.iter().for_each(|(alias, expr)| {
+                        let evaluated_val = expr.evaluate(&v_as_tuple, ctx).into_owned();
+                        if evaluated_val == Missing {
+                            evaluated_values.insert(alias.as_str(), Null);
+                        } else {
+                            evaluated_values.insert(alias.as_str(), evaluated_val);
+                        }
+                    });
+                    groups
+                        .entry(evaluated_values)
+                        .or_insert(vec![])
+                        .push(Value::Tuple(Box::new(v_as_tuple)));
+                }
+
+                let bag = groups
+                    .into_iter()
+                    .map(|(k, v)| match group_as_alias {
+                        None => Value::from(k),
```

Comment:
Eliding the values here and using just the group key will be insufficient when we add aggregations functions, as they will have nothing to aggregate. But I suppose this can be extended as part of that work.

Reply:
Yeah, I'm realizing that as I'm working through aggregations currently. Removing the values when a `GROUP AS` grouping variable is not specified won't allow for proper access when an aggregation is called over a variable not in the `GROUP BY` key list. E.g. `SELECT SUM(t.b) AS s FROM <<{'a': 1, 'b': 2},{'a': 1, 'b': 3}>> AS t GROUP BY t.a` won't work with how the current `GROUP BY` implementation removes the values.

Still exploring other ways to handle such cases. I added a TODO in the latest commit to track better.